ghcide/src/Development/IDE/Core/RuleTypes.hs

```diff
@@ -90,13 +93,38 @@ newtype ImportMap = ImportMap
   } deriving stock Show
     deriving newtype NFData
 
+data SpliceInfo = SpliceInfo
+    { exprSplices :: [(LHsExpr GhcTc, LHsExpr GhcPs)]
+    , patSplices :: [(LHsExpr GhcTc, LPat GhcPs)]
+    , typeSplices :: [(LHsExpr GhcTc, LHsType GhcPs)]
+    , declSplices :: [(LHsExpr GhcTc, [LHsDecl GhcPs])]
+    , awSplices :: [(LHsExpr GhcTc, Serialized)]
+    }
+
+instance Semigroup SpliceInfo where
+    SpliceInfo e p t d aw <> SpliceInfo e' p' t' d' aw' =
+        SpliceInfo
+            (e <> e')
+            (p <> p')
+            (t <> t')
+            (d <> d')
+            (aw <> aw')
+
+instance Monoid SpliceInfo where
+    mempty = SpliceInfo mempty mempty mempty mempty mempty
+
+instance NFData SpliceInfo where
+  rnf SpliceInfo{..} =
+     exprSplices `seq` patSplices `seq` typeSplices `seq` declSplices `seq` ()
```

Comment:
Use `rnf` to fully evaluate this.

Reply:
`HsExpr` and his friends don't come with NFData instance, so I chose to reduce them just to WHNF, as it seems there already exist such cases just reduces fields just to whnf. Are there any wrappers to generically traverse them?

Reply:
You can add `NFData` instances for `HsExpr` etc. in the Orphans module. The implementation for those can just be `seq`. That way, at least the spine of the list is forced. Alternatively, you can use `liftRnf seq`.